[pkg/ottl] Change grammar to support expressing statements context via path names

edmocosta opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

This PR is part of the https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017, and introduces the necessary OTTL Grammar/Parser changes to make it possible for statements express their context via path names.

This first iteration changes the grammar to parse the first `path` segment as the `Context` name, for example, the path `foo.bar.field` would be parsed as the following AST example: `{Context: "foo", Fields: [{Name: "bar"}, {Name: "field"}]}`, instead of `{Fields: [{Name: "foo"},{Name: "bar"}, {Name: "field"}]}`.

To make it non-breaking to all components during the transition time and development, this PR also changes the `ottl.Parser[k].newPath` function to, by default, fall this new behavior back to the previous one, using the grammar's parsed `Context` value as the first path segment (`Fields[0]`). 

Two new `Parser[K]` options were added, `WithPathContextNames` and `WithPathContextNameValidation`, The first option will be used to enable the statements context support, parsing the first path segment as Context when the value maches any configured context names, or falling it back otherwise. The second option will be used to enable the validation and turning off the backward compatible mode, raising an error when paths have no context prefix, or when it's invalid.

For more details, please check the https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017 discussion out.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017

**Testing:** 
- Unit tests

**Documentation:**
- No documentation changes were made at this point.

edmocosta committed:
Change the OTTL grammar to support expressing statements context via path names

edmocosta committed:
Merge branch 'main' of https://github.com/edmocosta/opentelemetry-collector-contrib into ottl_statements_context_change_grammar

edmocosta committed:
Fix failing test after branch update, adding more tests cases and nitpicks

TylerHelmuth commented on pkg/ottl/parser.go:
```diff
@@ -100,6 +102,32 @@ func WithEnumParser[K any](parser EnumParser) Option[K] {
+// WithPathContextNameValidation turns the validation mode on/off for the Path's contexts.
+// When enabled, and WithPathContextNames is configured, it requires all Path's to be prefixed
+// with a valid context name, matching at least one configured WithPathContextNames value.
+func WithPathContextNameValidation[K any](enabled bool) Option[K] {
```
Can we simplify the implementation by adding only `WithPathContextNames`? Then in `parsePathContext` if `pathContextNames` is set we assume the new logic should be used and if it is not set then the old logic should be used.

edmocosta replied:
I've added this option to help during the transition, the idea was later changing all `ottl/contexts/*` `NewParser(..)` functions adding their respective `WithPathContextNames` values, and modifying them to handle the path's context by default, without breaking or enforcing it yet.

This option makes it possible for a single Parser instance to support both logics, The `transformprocessor` for example, would leverage it to support the structured and flat configuration modes, without the need of modifying the existing statements or using multiple parsers instances with and without that configuration set.

I think its' feasible to drop this option from the Parser as suggested and make this code simpler. Later we can delegate the logic to enable the new behaviour to the ottlcontexts, possibly by adding a new `Option` that would configure the underline parser's `WithPathContextNames`, so components would be able to optionally enable it.

To drop it and avoid the need of multiple parsers instances w/ different configs (`transformprocessor` transition), we would also need to to build the statements rewrite utility not only for displaying user's the new statement syntax, but also for passing the modified value to the `ParseStatements` func.

WDYT?

TylerHelmuth replied:
> I think its' feasible to drop this option from the Parser as suggested and make this code simpler. Later we can delegate the logic to enable the new behaviour to the ottlcontexts, possibly by adding a new Option that would configure the underline parser's WithPathContextNames, so components would be able to optionally enable it.

Lets try this approach. If it doesn't work out we can always re-add this ottl parser option and implement things this way, but for now I'd like to reduce complexity in the core ottl package whenever possible.

edmocosta replied:
Done!

edmocosta committed:
Removed `WithPathContextNameValidation` option

edmocosta committed:
Fix test to use assert.Len (testifylint)

TylerHelmuth commented on pkg/ottl/functions.go:
```diff
@@ -64,10 +77,51 @@ func newPath[K any](fields []field) (*basePath[K], error) {
+			return "", path.Fields, fmt.Errorf(`context "%s" from path "%s" is not valid, it must be replaced by one of: %s`, path.Context, buildOriginalText(path), p.buildPathContextNamesText(""))
+		}
+		return path.Context, path.Fields, nil
+	} else if hasPathContextNames {
```
We can remove this else.

TylerHelmuth reviewed (commented):
@edmocosta I like the way this is looking.

For big refactors like this is gonna end up being we typically create a draft PR with all the changes implemented as if we were going to merge in a breaking change. Can you create a draft PR like that for this work? I'm interested to see the final picture for what we're trying to achieve before we start merging things. Don't worry about unit tests, I just want to see who the API and implementation shakes out.

edmocosta commented:
> @edmocosta I like the way this is looking.
> 
> For big refactors like this is gonna end up being we typically create a draft PR with all the changes implemented as if we were going to merge in a breaking change. Can you create a draft PR like that for this work? I'm interested to see the final picture for what we're trying to achieve before we start merging things. Don't worry about unit tests, I just want to see who the API and implementation shakes out.

Sure thing! I'll create a draft PR with all the pieces together soon! Thanks @TylerHelmuth!

edmocosta commented:
Hi @TylerHelmuth! here's the draft PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050
It includes all the changes we've been discussing. It certainly needs more work/discussions, including extra changes, adding tests, docs, code style, etc. It's a working draft implementation, so If you want you can try it out locally.

TylerHelmuth commented on .chloggen/ottl_statements_context_change_grammar.yaml:
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
```
```suggestion
change_type: breaking
```

Since we are adding a method to the interface we need to mark this as breaking.

edmocosta replied:
Oh that's true, good catch!

TylerHelmuth commented on .chloggen/ottl_statements_context_change_grammar.yaml:
```diff
@@ -0,0 +1,27 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
The `ottl.Path` interface requires a new method: `Context() string`

TylerHelmuth reviewed (commented):
@edmocosta thanks for quickly writing up that draft. There is definitely a lot of complexity we're adding with this change, but I believe the portability and readability of:

```
trace_statements:
  - set(span.status.code, 1) where span.attributes["http.path"] == "/health"
```

is really nice. Take a look at the remaining comments and then we can get this merged and move on to the next step.

edmocosta committed:
Applied PR suggestions, changed pathContextNames set type, and added "possibly" on the error message

edmocosta commented:
@TylerHelmuth, I've addressed the suggestions, plus two nitpicks that I changed on the draft PR:
- Changed the [pathContextNames](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34875/files#diff-79e887c0c48e52e9341081ec2f052ca7b3ecc9ec98dd657e7189bc559c9a80fbR70) type from `map[string]bool` to `map[string]struct{}`
- Added the `possibly` word on the error message [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34875/files#diff-7d00262f5fd3b4cb4b569750e059f83615378f288c84266f86b59b3747143f71R99). IMO, It's clear specially when the statement is invalid and only contains ambiguous context name, for example `set(resource.attributes["foo"], invalid.context)`
